app/models/manageiq/providers/openstack/infra_manager/refresh_parser.rb

```diff
@@ -289,46 +289,30 @@ def lookup_connection_state(power_state_input)
       def get_clusters
         # This counts with hosts being already collected
         hosts = @data.fetch_path(:hosts)
-        clusters = infer_clusters_from_hosts(hosts)
-
+        clusters, cluster_host_mapping = get_clusters_and_host_mapping
         process_collection(clusters, :clusters) { |cluster| parse_cluster(cluster) }
-        set_relationship_on_hosts(hosts)
-      end
-
-      def host_type(host)
-        host_type = host[:name].scan(/\((.*?)\)/).first
-        host_type.first if host_type
-      end
-
-      def cluster_name_for_host(host)
-        # TODO(lsmola) name and uid should also contain stack name, add this after the patch that saves resources
-        # is merged. Adding Overcloud by hard now.
-        host_type = host_type(host)
-        "overcloud #{host_type}"
-      end
 
-      def cluster_index_for_host(host)
-        # TODO(lsmola) name and uid should also contain stack name, add this after the patch that saves resources
-        # is merged. Adding Overcloud by hard now.
-        host_type = host_type(host)
-        "overcloud__#{host_type}"
+        set_relationship_on_hosts(hosts, cluster_host_mapping)
       end
 
-      def infer_clusters_from_hosts(hosts)
-        # We will create Cluster per Stack Host type. This way we can work with the same host types together
-        # as a group, e.g. all Compute hosts all Object Storage hosts, etc.
+      def get_clusters_and_host_mapping
         clusters = []
-        hosts.each do |host|
-          host_type = host_type(host)
-          # skip the non-provisoned hosts
-          next unless host_type
-
-          name = cluster_name_for_host(host)
-          uid = cluster_index_for_host(host)
+        cluster_host_mapping = {}
+        @data_index.fetch_path(:orchestration_stacks).each_pair do |_key, stack|
+          uid = stack.fetch_path(:parent, :ems_ref)
+          next unless uid
+
+          nova_servers_present = false
+          stack[:resources].each do |resource|
+            if resource[:resource_category] == 'OS::Nova::Server'
+              cluster_host_mapping[resource[:physical_resource]] = uid
+              nova_servers_present = true
```

Comment:
Should there be a `break` here?

Or, maybe even just move line 313 into the loop and break when the `'OS::Nova::Server'` is found.

```ruby
stack[:resources].each do |resource|
  if resource[:resource_category] == 'OS::Nova::Server'
    cluster_host_mapping[resource[:physical_resource]] = uid
    clusters << {:name => stack[:parent][:name], :uid => uid}
    break
  end
end
```

I guess the question is how many `'OS::Nova::Server'` resource categories will there be?  If there's only one, you could also just `detect` the `'OS::Nova::Server'` resource.

I'm just not a fan of saving some status from the loop to use later.